File: backend/app/services/calculation_engine.py
import logging
from typing import List, Dict, Any, Tuple
from motor.motor_asyncio import AsyncIOMotorDatabase

logger = logging.getLogger("app.services.calculation_engine")

# Conversion table to base units (kWh, litre, m3, kg)
CONVERSION_TABLE = {
    # Energy
    ("kwh", "mwh"): 1000.0,
    ("mwh", "kwh"): 0.001,
    ("litre", "gallon"): 3.78541,
    ("gallon", "litre"): 0.264172,
    ("litre", "l"): 1.0,
    ("l", "litre"): 1.0,
    # Material / Waste
    ("kg", "lb"): 0.453592,
    ("lb", "kg"): 2.20462,
    ("kg", "ton"): 1000.0,
    ("ton", "kg"): 0.001,
    ("kg", "tonne"): 1000.0,
    ("tonne", "kg"): 0.001
}

def normalize_unit_and_qty(input_unit: str, target_unit: str, quantity: float) -> Tuple[float, bool]:
    u_in = input_unit.lower().strip()
    u_tgt = target_unit.lower().strip()

    if u_in == u_tgt:
        return quantity, True

    # Check direct conversion
    key = (u_tgt, u_in)
    if key in CONVERSION_TABLE:
        multiplier = CONVERSION_TABLE[key]
        return quantity * multiplier, True

    return quantity, False

async def get_emission_factor(db: AsyncIOMotorDatabase, category: str, subtype: str, unit: str) -> Dict[str, Any]:
    cat = category.lower().strip()
    sub = subtype.lower().strip()
    u = unit.lower().strip()

    # Exact match search
    factor_doc = await db.emission_factors.find_one({
        "category": cat,
        "subtype": sub,
        "unit": u
    })
    if factor_doc:
        return factor_doc

    # Subtype fallback match with unit conversion check
    cursor = db.emission_factors.find({"category": cat, "subtype": sub})
    factors = await cursor.to_list(length=50)

    for ef in factors:
        target_unit = ef.get("unit", "")
        _, can_convert = normalize_unit_and_qty(u, target_unit, 1.0)
        if can_convert:
            return ef

    return None

async def calculate_assessment_co2e(db: AsyncIOMotorDatabase, assessment_id: str) -> Dict[str, Any]:
    """
    The ONLY authoritative function allowed to calculate computed_co2e for ProcessInputs and total_co2e for Assessment.
    Returns summary dict with computed total_co2e, category totals, line items, and unsupported inputs.
    """
    cursor = db.process_inputs.find({"assessment_id": assessment_id})
    input_items = await cursor.to_list(length=500)

    total_co2e_kg = 0.0
    line_details = []
    unsupported = []

    for item in input_items:
        item_id = str(item["_id"])
        category = item["category"]
        subtype = item["subtype"]
        quantity = float(item["quantity"])
        unit = item["unit"]

        ef_doc = await get_emission_factor(db, category, subtype, unit)

        if not ef_doc:
            unsupported.append({
                "input_id": item_id,
                "category": category,
                "subtype": subtype,
                "quantity": quantity,
                "unit": unit,
                "reason": f"No emission factor found for {category}/{subtype} in unit '{unit}'"
            })
            # Exclude from total, keep computed_co2e = 0.0
            await db.process_inputs.update_one(
                {"_id": item["_id"]},
                {"$set": {"computed_co2e": 0.0, "emission_factor_id": None}}
            )
            continue

        target_unit = ef_doc.get("unit", unit)
        factor_val = float(ef_doc.get("emission_factor", 0.0))
        factor_source = ef_doc.get("source", "Sourced Dataset")
        ef_id = str(ef_doc["_id"])

        normalized_qty, _ = normalize_unit_and_qty(unit, target_unit, quantity)
        line_co2e_kg = normalized_qty * factor_val
        line_co2e_tonnes = line_co2e_kg / 1000.0

        total_co2e_kg += line_co2e_kg

        # Update input line in DB
        await db.process_inputs.update_one(
            {"_id": item["_id"]},
            {"$set": {
                "computed_co2e": round(line_co2e_tonnes, 4),
                "emission_factor_id": ef_id
            }}
        )

        line_details.append({
            "id": item_id,
            "category": category,
            "subtype": subtype,
            "quantity": quantity,
            "unit": unit,
            "emission_factor_used": factor_val,
            "factor_source": factor_source,
            "computed_co2e": round(line_co2e_tonnes, 4),
            "line_co2e_kg": line_co2e_kg
        })

    total_co2e_tonnes = round(total_co2e_kg / 1000.0, 4)

    # Compute contribution percentages
    category_kg = {}
    for line in line_details:
        cat = line["category"]
        category_kg[cat] = category_kg.get(cat, 0.0) + line["line_co2e_kg"]
        
        line_pct = (line["line_co2e_kg"] / total_co2e_kg * 100.0) if total_co2e_kg > 0 else 0.0
        line["contribution_pct"] = round(line_pct, 2)

    category_totals = []
    for cat, cat_kg in category_kg.items():
        cat_tonnes = round(cat_kg / 1000.0, 4)
        cat_pct = round((cat_kg / total_co2e_kg * 100.0), 2) if total_co2e_kg > 0 else 0.0
        category_totals.append({
            "category": cat,
            "total_co2e": cat_tonnes,
            "pct": cat_pct
        })

    # Update assessment document
    await db.assessments.update_one(
        {"_id": assessment_id},
        {"$set": {"total_co2e": total_co2e_tonnes}}
    )
    # Also update if ObjectId
    from bson import ObjectId
    if ObjectId.is_valid(assessment_id):
        await db.assessments.update_one(
            {"_id": ObjectId(assessment_id)},
            {"$set": {"total_co2e": total_co2e_tonnes}}
        )

    return {
        "total_co2e": total_co2e_tonnes,
        "category_totals": category_totals,
        "line_items": line_details,
        "unsupported_inputs_count": len(unsupported),
        "unsupported_inputs": unsupported
    }

File: backend/src/services/calculationEngine.js
const EmissionFactor = require('../models/EmissionFactor');
const ProcessInput = require('../models/ProcessInput');
const Assessment = require('../models/Assessment');
const AssessmentHistory = require('../models/AssessmentHistory');

// Unit conversion table: converts from input unit to canonical factor unit
// Returns multiplier to apply to quantity BEFORE multiplying by emission factor
const UNIT_CONVERSIONS = {
  // Energy volume
  mwh: { to: 'kwh', factor: 1000 },
  kwh: { to: 'kwh', factor: 1 },
  // Liquid volume
  l: { to: 'litre', factor: 1 },
  litre: { to: 'litre', factor: 1 },
  liter: { to: 'litre', factor: 1 },
  gallon: { to: 'litre', factor: 3.78541 },
  gal: { to: 'litre', factor: 3.78541 },
  // Mass
  kg: { to: 'kg', factor: 1 },
  kilogram: { to: 'kg', factor: 1 },
  lb: { to: 'kg', factor: 0.453592 },
  pound: { to: 'kg', factor: 0.453592 },
  tonne: { to: 'kg', factor: 1000 },
  ton: { to: 'kg', factor: 1000 },
  mt: { to: 'kg', factor: 1000 },
  // Volume (gas/liquid)
  m3: { to: 'm3', factor: 1 },
  // Water (kilolitre = 1000 litres)
  kl: { to: 'kl', factor: 1 },
  kilolitre: { to: 'kl', factor: 1 },
  kiloliter: { to: 'kl', factor: 1 },
};

// Maps canonical emission factor units to lookup keys in emission factors collection
const EF_UNIT_MAP = {
  'kwh': ['kwh', 'kWh'],
  'litre': ['litre', 'l', 'liter'],
  'kg': ['kg'],
  'm3': ['m3'],
  'kl': ['kl'],
};

/**
 * Normalize a unit string to a canonical form
 */
function normalizeUnit(unit) {
  return (unit || '').toLowerCase().trim();
}

/**
 * Look up emission factor for a given category/subtype/unit combination
 * Returns { factor, factorDoc } or null
 */
async function lookupEmissionFactor(category, subtype, unit) {
  const normalUnit = normalizeUnit(unit);

  // Try direct subtype match first
  const factor = await EmissionFactor.findOne({
    category,
    subtype: { $regex: new RegExp(`^${subtype}$`, 'i') },
  });

  if (factor) return factor;

  // If not found, return null
  return null;
}

/**
 * Convert quantity from input unit to emission factor's canonical unit
 * Returns { convertedQuantity, canonicalUnit } or null if conversion unsupported
 */
function convertUnit(quantity, inputUnit, targetUnit) {
  const normInput = normalizeUnit(inputUnit);
  const normTarget = normalizeUnit(targetUnit);

  if (normInput === normTarget) return { convertedQuantity: quantity, canonicalUnit: targetUnit };

  // Check conversion table
  const conv = UNIT_CONVERSIONS[normInput];
  if (conv && conv.to === normTarget) {
    return { convertedQuantity: quantity * conv.factor, canonicalUnit: normTarget };
  }

  // Try reverse
  for (const [key, val] of Object.entries(UNIT_CONVERSIONS)) {
    if (key === normInput && val.to === normTarget) {
      return { convertedQuantity: quantity * val.factor, canonicalUnit: normTarget };
    }
  }

  return null; // unsupported conversion
}

/**
 * Main calculation engine — runs over all ProcessInput rows for an assessment
 * Sets computed_co2e on each input and total_co2e on the assessment
 */
async function calculateEmissions(assessmentId) {
  const assessment = await Assessment.findById(assessmentId);
  if (!assessment) throw new Error('Assessment not found');

  const inputs = await ProcessInput.find({ assessment_id: assessmentId });
  if (inputs.length === 0) {
    throw { code: 'VALIDATION_ERROR', message: 'No process inputs to calculate', status: 400 };
  }

  let totalCo2eKg = 0;
  const unsupportedInputs = [];
  const lineItems = [];

  for (const input of inputs) {
    const efDoc = await lookupEmissionFactor(input.category, input.subtype, input.unit);

    if (!efDoc) {
      // Mark as unsupported — still stored, flagged
      await ProcessInput.findByIdAndUpdate(input._id, { computed_co2e: null, unsupported: true, emission_factor_id: null });
      unsupportedInputs.push({
        input_id: input._id,
        category: input.category,
        subtype: input.subtype,
        unit: input.unit,
        reason: 'no matching emission factor',
      });
      continue;
    }

    // Convert unit
    const convResult = convertUnit(input.quantity, input.unit, efDoc.unit);
    let convertedQty = input.quantity;
    if (convResult) {
      convertedQty = convResult.convertedQuantity;
    }

    // CO2e in kg
    const co2eKg = convertedQty * efDoc.emission_factor;
    // Convert to tonnes for storage
    const co2eTonnes = co2eKg / 1000;

    await ProcessInput.findByIdAndUpdate(input._id, {
      computed_co2e: co2eTonnes,
      emission_factor_id: efDoc._id,
      unsupported: false,
    });

    totalCo2eKg += co2eKg;

    lineItems.push({
      id: input._id,
      category: input.category,
      subtype: input.subtype,
      quantity: input.quantity,
      unit: input.unit,
      emission_factor: efDoc.emission_factor,
      factor_unit: efDoc.factor_unit,
      factor_source: efDoc.source,
      co2e: parseFloat(co2eTonnes.toFixed(4)),
    });
  }

  const totalCo2eTonnes = parseFloat((totalCo2eKg / 1000).toFixed(4));

  // Update assessment
  await Assessment.findByIdAndUpdate(assessmentId, {
    total_co2e: totalCo2eTonnes,
    status: 'complete',
    completed_at: new Date(),
  });

  // Record in history
  await AssessmentHistory.create({
    facility_id: assessment.facility_id,
    assessment_id: assessmentId,
    total_co2e: totalCo2eTonnes,
    recorded_at: new Date(),
  });

  return {
    total_co2e: totalCo2eTonnes,
    unsupported_inputs: unsupportedInputs,
    unsupported_inputs_count: unsupportedInputs.length,
    line_items: lineItems,
  };
}

/**
 * Build summary data (for GET /assessments/:id/summary)
 */
async function buildSummary(assessmentId) {
  const assessment = await Assessment.findById(assessmentId);
  if (!assessment) throw { code: 'NOT_FOUND', message: 'Assessment not found', status: 404 };
  if (assessment.status !== 'complete') {
    throw { code: 'ASSESSMENT_NOT_CALCULATED', message: 'Assessment has not been calculated yet', status: 400 };
  }

  const inputs = await ProcessInput.find({ assessment_id: assessmentId });
  const categoryTotals = { energy: 0, material: 0, waste: 0 };
  const lineItems = [];
  let unsupportedCount = 0;

  for (const input of inputs) {
    if (input.unsupported || input.computed_co2e === null) {
      unsupportedCount++;
      continue;
    }

    categoryTotals[input.category] = (categoryTotals[input.category] || 0) + input.computed_co2e;

    const efDoc = input.emission_factor_id
      ? await EmissionFactor.findById(input.emission_factor_id)
      : null;

    lineItems.push({
      id: input._id,
      category: input.category,
      subtype: input.subtype,
      quantity: input.quantity,
      unit: input.unit,
      emission_factor: efDoc ? efDoc.emission_factor : null,
      factor_unit: efDoc ? efDoc.factor_unit : null,
      factor_source: efDoc ? efDoc.source : null,
      co2e: parseFloat((input.computed_co2e || 0).toFixed(4)),
      pct_contribution: assessment.total_co2e > 0
        ? parseFloat((input.computed_co2e / assessment.total_co2e).toFixed(4))
        : 0,
    });
  }

  // Round category totals
  for (const k of Object.keys(categoryTotals)) {
    categoryTotals[k] = parseFloat(categoryTotals[k].toFixed(4));
  }

  return {
    total_co2e: assessment.total_co2e,
    category_totals: categoryTotals,
    line_items: lineItems,
    unsupported_inputs_count: unsupportedCount,
  };
}

module.exports = { calculateEmissions, buildSummary, lookupEmissionFactor, normalizeUnit };
